src/main/java/org/folio/consortia/service/impl/PublicationServiceImpl.java

```diff
@@ -0,0 +1,217 @@
+import org.folio.consortia.service.TenantService;
+import org.folio.consortia.service.UserTenantService;
+import org.folio.spring.FolioExecutionContext;
+import org.folio.spring.FolioModuleMetadata;
+import org.folio.spring.scope.FolioExecutionContextSetter;
+import org.springframework.core.task.TaskExecutor;
+import org.springframework.http.HttpMethod;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.stereotype.Service;
+import org.springframework.web.client.HttpClientErrorException;
+
+import com.fasterxml.jackson.core.JsonProcessingException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+
+import lombok.RequiredArgsConstructor;
+import lombok.SneakyThrows;
+import lombok.extern.log4j.Log4j2;
+
+@Service
+@Log4j2
+@RequiredArgsConstructor
+public class PublicationServiceImpl implements PublicationService {
+  private final TenantService tenantService;
+  private final UserTenantService userTenantService;
+  private final FolioExecutionContext folioExecutionContext;
+  private final FolioModuleMetadata folioModuleMetadata;
+  private final HttpRequestService httpRequestService;
+  private final TaskExecutor asyncTaskExecutor;
+
+  private final PublicationStatusRepository publicationStatusRepository;
+  private final PublicationTenantRequestRepository publicationTenantRequestRepository;
+  private final ObjectMapper objectMapper;
+  private static final int MAX_ACTIVE_THREADS = 5;
+
+  @Override
+  @SneakyThrows
+  public PublicationResponse publishRequest(UUID consortiumId, PublicationRequest publicationRequest) {
+    validatePublicationRequest(consortiumId, publicationRequest, folioExecutionContext);
+
+    PublicationStatusEntity createdPublicationEntity = persistPublicationRecord(publicationRequest.getTenants().size());
+
+    asyncTaskExecutor.execute(() -> processTenantRequests(publicationRequest, createdPublicationEntity));
+
+    return buildPublicationResponse(createdPublicationEntity.getId());
+  }
+  void processTenantRequests(PublicationRequest publicationRequest, PublicationStatusEntity createdPublicationEntity) {
+    List<CompletableFuture<PublicationTenantRequestEntity>> futures = new ArrayList<>();
+    Semaphore semaphore = new Semaphore(MAX_ACTIVE_THREADS);
+
+    for (String tenantId : publicationRequest.getTenants()) {
+      try {
+        PublicationTenantRequestEntity ptrEntity = buildPublicationRequestEntity(publicationRequest, createdPublicationEntity, tenantId);
+
+        var savedPublicationTenantRequest = savePublicationTenantRequest(ptrEntity, folioExecutionContext);
+
+        semaphore.acquire();
+        var future = executeAsyncHttpRequest(publicationRequest, tenantId, folioExecutionContext)
+          .whenComplete((ok, err) -> semaphore.release())
+          .handle((response, t) -> updatePublicationTenantRequest(response, t, savedPublicationTenantRequest, folioExecutionContext));
```

Comment:
Not sure if the value of injected folioExecutionContext would always be expected one for async invocations. Here's the scope configurer https://github.com/folio-org/folio-spring-support/blob/master/folio-spring-base/src/main/java/org/folio/spring/scope/FolioExecutionScopeConfigurer.java